In [1]:
#call this only once for a kernel startup
from __future__ import absolute_import, division, print_function

import tensorflow as tf
# BATCH_SIZE = 32
seed = 12
import sys
orig_stdout = sys.stdout
NoOfClasses = 2
dataType = 'discrete'
sys.stdout = orig_stdout

In [2]:
import numpy as np
sys.path.append("../")
import model_only

gold_labels_dev = np.load("data/true_labels_dev.npy")
gold_labels_test = np.load("data/true_labels_test.npy")

unique, counts = np.unique(gold_labels_dev, return_counts=True)
print(dict(zip(unique, counts)))

unique, counts = np.unique(gold_labels_test, return_counts=True)
print(dict(zip(unique, counts)))

print(len(gold_labels_dev),len(gold_labels_test))

{0: 2625, 1: 189}
{0: 2484, 1: 218}
2814 2702


In [3]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from collections import defaultdict
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_recall_fscore_support
from sklearn.metrics import precision_recall_curve
from sklearn.metrics import average_precision_score
from sklearn.metrics import roc_auc_score

def draw2DArray(a):
    fig = plt.figure(figsize=(6, 3.2))
    ax = fig.add_subplot(111)
    ax.set_title('colorMap')
    plt.imshow(np.array(a))
    ax.set_aspect('equal')
    cax = fig.add_axes([0.12, 0.1, 0.78, 0.8])
    cax.get_xaxis().set_visible(False)
    cax.get_yaxis().set_visible(False)
    cax.patch.set_alpha(0)
    cax.set_frame_on(False)
    plt.colorbar(orientation='vertical')
    plt.show()
    
      
def report2dict(cr):
    # Parse rows
    tmp = list()
    for row in cr.split("\n"):
        parsed_row = [x for x in row.split("  ") if len(x) > 0]
        if len(parsed_row) > 0:
            tmp.append(parsed_row)
    
    # Store in dictionary
    measures = tmp[0]

    D_class_data = defaultdict(dict)
    for row in tmp[1:]:
        class_label = row[0]
        for j, m in enumerate(measures):
            D_class_data[class_label][m.strip()] = float(row[j + 1].strip())
    return pd.DataFrame(D_class_data).T

def predictAndPrint(pl):
    print("acc",accuracy_score(gold_labels_dev,pl))
    print(precision_recall_fscore_support(gold_labels_test,pl,average='binary'))
    print(confusion_matrix(gold_labels_dev,pl))
#     draw2DArray(confusion_matrix(gold_labels_dev,pl))
#     return report2dict(classification_report(gold_labels_dev, pl))# target_names=class_names))
    


def drawPRcurve(y_test,y_score,it_no):
    
    fig = plt.figure()
    splt = fig.add_subplot(111)
    

    precision, recall, _ = precision_recall_curve(y_test, y_score,pos_label=1)

    splt.step(recall, precision, color='b', alpha=0.2,
             where='post')
    splt.fill_between(recall, precision, step='post', alpha=0.2,
                     color='b')
    average_precision = average_precision_score(y_test, y_score)
    plt.xlabel('Recall')
    plt.ylabel('Precision')
    plt.ylim([0.0, 1.05])
    plt.xlim([0.0, 1.05])
    plt.title('{0:d} Precision-Recall curve: AP={1:0.2f}'.format(it_no,
              average_precision))
  

In [4]:
if dataType == 'merged':
    LF_l = [
        1,1,1,1,1,-1,1,-1,-1,-1
    ]
    dscr = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0]
    def merge(a,b):
        c = []
        for i in range(len(a)):
            ci = []
            ci_l = a[i,0,:].tolist()+b[i,0,:].tolist()
            ci_s = a[i,1,:].tolist()+b[i,1,:].tolist()
            ci.append(ci_l)
            ci.append(ci_s)
            c.append(ci)
        return c
    import numpy as np
    dev_L_S_s = np.load("data/dev_L_S_smooth.npy")
    test_L_S_s = np.load("data/test_L_S_smooth.npy")
    train_L_S_s = np.load("data/train_L_S_smooth.npy")

    dev_L_S_d = np.load("data/dev_L_S_discrete.npy")
    test_L_S_d = np.load("data/test_L_S_discrete.npy")
    train_L_S_d = np.load("data/train_L_S_discrete.npy")

    dev_L_S = np.array(merge(dev_L_S_d,dev_L_S_s))
    train_L_S = np.array(merge(train_L_S_d,train_L_S_s))
    test_L_S = np.array(merge(test_L_S_d,test_L_S_s))

    LF_l = LF_l + LF_l
    NoOfLFs= len(LF_l)

    print(len(LF_l))

    print(dev_L_S.shape, test_L_S.shape, train_L_S.shape)
    print(train_L_S[100])

In [5]:
if dataType == 'discrete':
    LF_l = [
        1,1,1,1,1,-1,1,-1,-1,-1
    ]
    NoOfLFs= len(LF_l)
    dscr = np.ones(10)

    dev_L_S = np.load("data/dev_L_S_discrete.npy")
    test_L_S = np.load("data/test_L_S_discrete.npy")
    train_L_S = np.load("data/train_L_S_discrete.npy")

    print(len(gold_labels_dev),len(gold_labels_test))
    print(test_L_S.shape,dev_L_S.shape,train_L_S.shape)
    print(train_L_S[100])

2814 2702
(2702, 2, 10) (2814, 2, 10) (22276, 2, 10)
[[ 0.   0.   0.   0.   0.   0.   0.  -1.  -1.   0. ]
 [ 0.5  0.5  0.5  0.5  0.5  0.5  0.5  1.   1.   0.5]]


In [6]:
if dataType == 'cont':
    LF_l = [
        1,1,1,1,1,-1,1,-1,-1,-1
    ]
    dscr = np.array([1,1,0,0,1,1,0,0,0,0])
    NoOfLFs= len(LF_l)

    dev_L_S = np.load("data/dev_L_S_smooth.npy")
    test_L_S = np.load("data/test_L_S_smooth.npy")
    train_L_S = np.load("data/train_L_S_smooth.npy")

    print(len(gold_labels_dev),len(gold_labels_test))
    print(test_L_S.shape,dev_L_S.shape,train_L_S.shape)
    print(train_L_S[0:10])

In [7]:

discr = []
for j in range(len(LF_l)):
    temp = np.array(train_L_S[:,0,j]*np.array(train_L_S[:,1,j]))
    unique, counts = np.unique(temp, return_counts=True)
    temp_d = dict(zip(unique, counts))
    print(j, '\t', len(temp_d), temp_d if len(temp_d)==2 else '')
    discr.append(int(not(len(temp_d)-2)))
sys.stdout.write('[')
for x in range(len(discr)-1):
    sys.stdout.write(str(discr[x]) + ', ')
sys.stdout.write(str(discr[-1])+']')

print(len(LF_l))

sys.stdout = orig_stdout
print('test')

0 	 2 {0.0: 22243, 1.0: 33}
1 	 2 {0.0: 22096, 1.0: 180}
2 	 2 {0.0: 19945, 1.0: 2331}
3 	 2 {0.0: 20538, 1.0: 1738}
4 	 2 {0.0: 21904, 1.0: 372}
5 	 2 {-1.0: 13366, 0.0: 8910}
6 	 2 {0.0: 22261, 1.0: 15}
7 	 2 {-1.0: 2323, 0.0: 19953}
8 	 2 {-1.0: 1634, 0.0: 20642}
9 	 2 {-1.0: 208, 0.0: 22068}
[1, 1, 1, 1, 1, 1, 1, 1, 1, 1]10
test


In [8]:
#added new model
def train2(lr,ep,th,af,batch_size=32,LF_acc=None,LF_rec=None,pcl=np.array([-1,1],dtype=np.float64),norm=True,\
          smooth=True,penalty=0,p3k=3,alp=1,Gamma=1.0,debug=True, \
           r1=1.0, r2=1.0, a_pr=None, n_pr=None, user_a = None, isdiscrete = None, alpha_max = None, r_pr=None):
    
    ## lr : learning rate
    ## ep : no of epochs
    ## th : thetas initializer
    ## af : alphas initializer
    ## penalty : {1,2,3} use one of the three penalties, 0: no-penalty
    ## p3k : parameter for penalty-3 
    ## smooth : flag if smooth lfs are used 
    ## make sure smooth/discrete LF data is loaded into train_L_S and test_L_S
    ## pcl : all possible class labels  = [-1,1] for binary, 
    ##       np.arange(0,NoOfClasses) for multiclass
    ## alp : alpha parameter (to set a max value for alpha)
    ## norm : use normalization or not
    ## Gamma : penalty tuning parameter
    
    BATCH_SIZE = batch_size
    tf.reset_default_graph()

    seed = 12
    
    LF_k = []
    if(len(pcl)==2):
        for i in range(len(LF_l)):
            LF_k.append(int((LF_l[i]+1)/2))
    with tf.Graph().as_default():
        xtra = tf.Variable(0, trainable=False)
        gammaR = tf.constant(0, dtype=tf.float64)

        train_dataset = tf.data.Dataset.from_tensor_slices(train_L_S).batch(BATCH_SIZE)
        dev_dataset = tf.data.Dataset.from_tensor_slices(dev_L_S).batch(len(dev_L_S))
        test_dataset = tf.data.Dataset.from_tensor_slices(test_L_S).batch(len(test_L_S))

        iterator = tf.data.Iterator.from_structure(train_dataset.output_types, train_dataset.output_shapes)
        next_element = iterator.get_next()

        train_init_op = iterator.make_initializer(train_dataset)
        dev_init_op = iterator.make_initializer(dev_dataset)
        test_init_op = iterator.make_initializer(test_dataset)

        next_element = iterator.get_next()

        thetas,alphas = model_only.allocate_params(NoOfLFs, 2, penalty, th, af)
        
        LF_label = tf.convert_to_tensor(LF_k, dtype=tf.int32)
        k = tf.convert_to_tensor(LF_l, dtype=tf.float64)
        g = tf.convert_to_tensor(Gamma, dtype=tf.float64)
        
        if(penalty in [4,5,7,8,9,10,11, 44, 42, 41]):
            LF_a = tf.convert_to_tensor(LF_acc, dtype=tf.float64)
        
        if(penalty in [6,7,11,12,15]):
            LF_r = tf.convert_to_tensor(LF_rec, dtype=tf.float64)
        
        l,s =  tf.unstack(next_element,axis=1)

        if(smooth and penalty < 100):
            s_ = tf.maximum(tf.subtract(s,tf.minimum(alphas,alp)), 0)
            if(debug):
                print("s_",s_)

        def iskequalsy(v,s):
            out = tf.where(tf.equal(v,s),tf.ones_like(v),-tf.ones_like(v))
            if(debug):
                print("out",out)
            return out
#MAP
        if(penalty < 100):
            if(smooth):
                pout = tf.map_fn(lambda c: l*c*s_ ,pcl,name="pout")
            else:
                pout = tf.map_fn(lambda c: l*c ,pcl,name="pout")

            t =  tf.squeeze(thetas)                
                
            def ints(y): # called for y=1 and y=-1
                ky = iskequalsy(k,y)  # ky = 1 if k==y else -1
                if(debug):
                    print("ky",ky)
                out1 = alphas+((tf.exp((t*ky*(1-alphas)))-1)/(t*ky))
                if(debug):
                    print("intsy",out1)
                return out1                
                
            if(smooth):
                #smooth normalizer
                zy = tf.map_fn(lambda y: tf.reduce_prod(1+ints(y),axis=0),\
                               pcl,name="zy")
            else:
                #discrete normalizer
                zy = tf.map_fn(lambda y: tf.reduce_prod(1+tf.exp(t*iskequalsy(k,y)),axis=0),\
                               pcl,name="zy")            
#new model

        if(penalty > 200):

#             z_y_new = [tf.reduce_prod(1 + 1/(-1*k*thetas) * (tf.exp(-1*k*thetas) - tf.exp(-1*k*thetas*user_alphas)) ), \
#                    tf.reduce_prod(1 + 1/(1*k*thetas) * (tf.exp(1*k*thetas) - tf.exp(1*k*thetas*user_alphas)) ) ]
            numbins = tf.constant(10, dtype=tf.float64)
            #TODO: 2 binned integration over the s values to allow stable gradients.
            sbin_widths = (1-user_alphas)/numbins
            sbins = tf.einsum('i,j->ij', sbin_widths, tf.cast(tf.range(0,numbins+1), dtype=tf.float64))+tf.expand_dims(user_alphas,1)
            sbins = tf.transpose(sbins)
            
            def pot(s):
                if(penalty%10 == 1):
                    return thetas*s-alphas
                elif(penalty%10 == 2):
                    return thetas*tf.clip_by_value(s-alphas,0,1)
                elif(penalty%10 == 3):
                    return thetas*s
                return 0
                                       
            # over the y-s it is okay to use a map function, but not over the LFs.
            z_y_new = tf.map_fn(lambda y: tf.reduce_prod(1 + tf.reduce_sum(tf.exp(-y*k*pot(sbins)), axis=0)), tf.constant([1,-1], dtype=tf.float64))
#             z_y_new = tf.reduce_sum(tf.exp(-1*k*thetas*sbins), axis=0)*sbin_widths/(1-user_alphas)
# *sbin_widths

            Z_new = tf.reduce_sum(z_y_new)
            print("new z ", Z_new)

#p theta without exp [y x i]
            # snap s_ to the nearest bin. 
            s_ = tf.round((s_-user_alphas)*numbins/(1-user_alphas))*sbin_widths + user_alphas
            logp_theta_new_t = tf.map_fn(lambda y:  tf.reduce_sum(-y*l*pot(s_), axis=1), tf.constant([1,-1], dtype=tf.float64))


            loss_new_t = tf.reduce_logsumexp(logp_theta_new_t, axis=0) - tf.log(Z_new)
            loss_new = tf.negative(tf.reduce_sum(loss_new_t))
        
#marginals as softmax

            marginals_new = tf.expand_dims(tf.nn.softmax(logp_theta_new_t, axis=0), 2)

        if(penalty < 100):
            
            if(debug):
                print("pout",pout)    
    #MAP
            t_pout = tf.map_fn(lambda x: tf.matmul(x,thetas,transpose_b=True),pout,\
                               name="t_pout")

            if(debug):
                print("t_pout",t_pout)

        t =  tf.squeeze(thetas)
        if(debug):
            print("t",t)

        if(penalty < 100):
            marginals = tf.nn.softmax(t_pout,axis=0)
#         else:
#             marginals = marginals_new
        
            if(debug):
                print("zy",zy)
            logz = tf.log(tf.reduce_sum(zy,axis=0),name="logz")
            if(debug):
                print("logz",logz)
    ##         tf.summary.scalar('logz', logz)
            lsp = tf.reduce_logsumexp(t_pout,axis=0)
            if(debug):
                print("lsp",lsp)
##         tf.summary.scalar('lsp', tf.reduce_sum(lsp))
        
    
        if(penalty == 20 or penalty > 100):
            a_t = tf.convert_to_tensor(a_pr, dtype=tf.float64)
            n_t = tf.convert_to_tensor(n_pr, dtype=tf.float64)
            r_t = tf.convert_to_tensor(r_pr, dtype=tf.float64)
        def get_loss_20():
#NEW MODEL

            # unnormalized_marginals [numY]
            unnormalized_marginals = [tf.reduce_prod(tf.exp(-thetas*LF_l) + 1), \
                                      tf.reduce_prod(tf.exp(thetas*LF_l) + 1)]
            # same for continuous, just use the new formulation with sbins for summation, no intergration
            Z = tf.reduce_sum(unnormalized_marginals)

            # unnormalized_marginals_tiled: [numLF, numY]
            # per-LF-marginals: [numLFs]
            per_LF_marginals = tf.gather(unnormalized_marginals, LF_label)

#             return per_LF_marginals
            # Pθ(kj=y): [numLFs]  (computing Eq 14)                       
            per_LF_agreement_prob = tf.exp(thetas)/(1 + tf.exp(thetas))*per_LF_marginals/Z

#             p_thetas =  tf.expand_dims(prod__,1) * exp_terms/(1+exp_terms) / Z
            p_thetas = per_LF_agreement_prob

            ptheta_terms = a_t * n_t * tf.log(p_thetas) + (1-a_t) * n_t * tf.log(1-p_thetas)
                                                   
#-------------------------------

            return tf.negative(tf.reduce_sum(ptheta_terms)) # add other loss component

        loss_recall = 0
        if(penalty > 100):
            numYs=2
            loss_new, per_lf_prob, marginals, per_lf_recall, pots = model_only.model(numYs, k, l, s, thetas, alphas, isdiscrete, user_a, penalty,alpha_max)
            loss_precision = model_only.precision_loss(a_t, n_t, per_lf_prob, penalty)
            if r_t is not None and (penalty//10) % 10 == 3:
                loss_recall = model_only.recall_loss(r_t, n_t, per_lf_recall, isdiscrete)

#adding ne      
           
        prec_loss = tf.constant(0)
    
        if(not norm):
            print("unnormlized loss")
            loss = tf.negative(tf.reduce_sum(lsp  ))
        elif(penalty == 1):
            print("penalty1")
            loss = tf.negative(tf.reduce_sum(lsp  - logz  )) \
                      +(g*tf.reduce_sum(tf.maximum(tf.zeros_like(thetas),-thetas)))
        elif(penalty == 2):
            print("penalty2")
            loss = tf.negative(tf.reduce_sum(lsp  - logz  )) \
                     -(g*tf.minimum( tf.reduce_min(thetas),0.0))
            prec_loss = tf.constant(0)
        elif(penalty == 3):
            print("penalty3")
            loss = tf.negative(tf.reduce_sum(lsp  - logz  )) \
                     +(g*tf.reduce_sum(tf.log(1+tf.exp(-thetas-p3k))))
        

        elif(penalty == 4):
            print("precision penalty")
            prec_loss = tf.reduce_sum(tf.map_fn(lambda j: softplus_p(j),np.arange(NoOfLFs),\
                                             dtype=tf.float64))
            loss = tf.negative(tf.reduce_sum(lsp  - logz  )) \
                + (g* prec_loss)
        
        elif(penalty == 41):
            print("precision penalty")
            prec_loss, xtra = sfp()
            loss = tf.negative(tf.reduce_sum(lsp  - logz  )) \
                + (g * prec_loss)
#                 + (g*tf.reduce_sum(tf.map_fn(lambda j: softplus_p(j),np.arange(NoOfLFs),\
#                                              dtype=tf.float64)))
        
        elif(penalty == 42):
            print("precision penalty")
            temp_prloss, _ = sfpp()
            temp_uloss = tf.negative(tf.reduce_sum(lsp  - logz  ))
            r21 = tf.gradients(temp_uloss, thetas)
            r22 = tf.gradients(temp_prloss, thetas)
            xtra_tfg = tf.reduce_sum(tf.abs(r21[0][0]))/tf.reduce_sum(tf.abs(r22[0][0]))
#             gammaR = temp_uloss/temp_prloss
#             prec_loss, xtra = sfpp()
            prec_loss = tf.reduce_sum(temp_prloss)

            grad_ratio = 1

            xtra = [xtra_tfg]

            loss = temp_uloss \
                + (g * prec_loss)

        elif(penalty == 44):
            print("precision penalty")
            loss = tf.negative(tf.reduce_sum(lsp  - logz  )) \
                + (g*tf.reduce_sum(tf.map_fn(lambda j: softplus_pp(j),np.arange(NoOfLFs),\
                                             dtype=tf.float64)))
                
                
        elif(penalty == 5):
            print("precision log(softplus) penalty")
            loss = tf.negative(tf.reduce_sum(lsp  - logz  )) \
                + (g*tf.reduce_sum(tf.map_fn(lambda j: tf.log(softplus_p(j)),np.arange(NoOfLFs),\
                                             dtype=tf.float64)))
        elif(penalty == 6):
            print("recall penalty")
            loss = tf.negative(tf.reduce_sum(lsp  - logz  )) \
                + (g*tf.reduce_sum(tf.map_fn(lambda j: softplus_r(j),np.arange(NoOfLFs),\
                                             dtype=tf.float64)))
        elif(penalty == 7):
            print("precision and recall penalty")
            loss = tf.negative(tf.reduce_sum(lsp  - logz  )) \
                + (g*tf.reduce_sum(tf.map_fn(lambda j: softplus_p(j),np.arange(NoOfLFs),\
                                             dtype=tf.float64))) \
                + (g*tf.reduce_sum(tf.map_fn(lambda j: softplus_r(j),np.arange(NoOfLFs),\
                                             dtype=tf.float64)))
        elif(penalty == 8):
            print("precision and sign 1 penalty")
            loss = tf.negative(tf.reduce_sum(lsp  - logz  )) \
                + (g*tf.reduce_sum(tf.map_fn(lambda j: softplus_p(j),np.arange(NoOfLFs),\
                                             dtype=tf.float64))) \
                + (g*tf.reduce_sum(tf.maximum(tf.zeros_like(thetas),-thetas)))
        elif(penalty == 9):
            print("precision and sign 2 penalty")
            loss = tf.negative(tf.reduce_sum(lsp  - logz  )) \
                + (g*tf.reduce_sum(tf.map_fn(lambda j: softplus_p(j),np.arange(NoOfLFs),\
                                             dtype=tf.float64))) \
                - (g*tf.minimum( tf.reduce_min(thetas),0.0))
        elif(penalty == 10):
            print("precision and sign 3 penalty")
            loss = tf.negative(tf.reduce_sum(lsp  - logz  )) \
                + (g*tf.reduce_sum(tf.map_fn(lambda j: softplus_p(j),np.arange(NoOfLFs),\
                                             dtype=tf.float64))) \
                + (g*tf.reduce_sum(tf.log(1+tf.exp(-thetas-p3k))))
        elif(penalty == 11):
            print("precision and recall and sign 2 penalty")
            loss = tf.negative(tf.reduce_sum(lsp  - logz  )) \
                + (g*tf.reduce_sum(tf.map_fn(lambda j: softplus_p(j),np.arange(NoOfLFs),\
                                             dtype=tf.float64))) \
                + (g*tf.reduce_sum(tf.map_fn(lambda j: softplus_r(j),np.arange(NoOfLFs),\
                                             dtype=tf.float64))) \
                  - (g*tf.minimum( tf.reduce_min(thetas),0.0))
        elif(penalty == 12):
            print("recall and sign 2 penalty")
            loss = tf.negative(tf.reduce_sum(lsp  - logz  )) \
                + (g*tf.reduce_sum(tf.map_fn(lambda j: softplus_r(j),np.arange(NoOfLFs),\
                                             dtype=tf.float64))) \
                  - (g*tf.minimum( tf.reduce_min(thetas),0.0))
        elif(penalty == 15):
            print("equation 15 and sign 2 penalty")
            loss = tf.negative(tf.reduce_sum(lsp  - logz  )) \
                + (g*tf.reduce_sum(tf.map_fn(lambda j: softplus_r15(j),np.arange(NoOfLFs),\
                                             dtype=tf.float64))) \
                  - (g*tf.minimum( tf.reduce_min(thetas),0.0))



                    
        elif(penalty == 20):
            print("new model")
            prec_loss = get_loss_20()
            loss =  tf.negative(tf.reduce_sum(lsp  - logz  )) + prec_loss
                  
        elif(penalty >100 and penalty<111):
            sys.stdout = stdout_saved
            print("new model smooth ", penalty)
            # sys.stdout = open("trash", "w")            
            print("new model smooth ", penalty)
            loss = loss_new
#             loss, _ = smooth_new_prec()
            
                    
        elif(penalty >110):
            sys.stdout = stdout_saved
            print("new model smooth with constraints ", penalty)
            # sys.stdout = open("trash", "w")            
            print("new model smooth with constraints ", penalty)
            loss = loss_new + loss_precision + loss_recall
#             loss_un, loss_pr = smooth_new_prec()
#             loss = loss_un + loss_pr

        else:
            print("normalized loss")
            loss = tf.negative(tf.reduce_sum(lsp  - logz  ))
            prec_loss = tf.constant(0)
       
        if(debug):
            print("loss",loss)

#         marginals = tf.nn.softmax(t_pout,axis=0)

        if(debug):
            print("marginals",marginals)
        predict = tf.argmax(marginals,axis=0)


        train_step = tf.train.AdamOptimizer(lr).minimize(loss)#, var_list=[thetas, alphas]) 


        init_g = tf.global_variables_initializer()
        init_l = tf.local_variables_initializer()
        
        dev_high = np.zeros(3)
        test_high = np.zeros(3)
        
        with tf.Session() as sess:
            sess.run(init_g)
            sess.run(init_l)

            flg = False
            # Initialize an iterator over the training dataset.
            for en in range(ep):
                sess.run(train_init_op)
#                 temprun = sess.run(next_element2)
#                 print('len next ',len(temprun), len(temprun[0]))
#                 print(temprun[0])
                tl = 0
                pll = 0
                unl = 0
                try:
                    it = 0
                    
                    grad_sum = 0.0
                    while True:
#                         print(en, it)

                        _,ls,ploss,lfProbs,t,al,p = sess.run([train_step,loss_new,loss_precision,per_lf_prob,thetas,alphas,pots])
                        if(penalty == 42):
                            un_loss = sess.run(temp_uloss)

                        tl = tl + ls
                        pll = pll + ploss
                        if(penalty == 42):
                            unl = unl + un_loss
                        it = it + 1
                        probs = np.exp(lfProbs)
                        if (np.amax(probs) > 1+1e-6):
                            print("Invalid probabilities=",probs)
                            
                        # print("various shapes=",p)
                        # print("Loss, prec_loss=", lfProbs)
                        # potsv, sv, lv, kv = sess.run([pots,s,l,k])
                        # print(t,al)
                        # print(sv,lv,kv)
                        # print(potsv)
#                         newptheta = sess.run(sbins)
#                         ztemp = sess.run(xyz)
                
#                         print(it*BATCH_SIZE)
                
                except tf.errors.OutOfRangeError:
                    pass
                opstring = ""
                print(en,"loss=",tl, " precision loss=",pll)
                opstring = opstring + str(en) + "," + str(tl) + ","
#                 print('temp__', temp__)
                
                print("ploss", pll)
                print("ratio", unl/pll)
#                 dloss = sess.run(dev_loss)
#                 print("dloss", dloss)
#                 print("expected ratio", xpp/(xpp+1))
                print("dev set")
                sess.run(dev_init_op)
##                 sm,a,t,m,pl = sess.run([summary_merged,alphas,thetas,marginals,predict])
#                 a,t,m,pl = sess.run([alphas,thetas,marginals,predict])
                a,t,m,pl,ddloss = sess.run([alphas,thetas,marginals,predict,loss])
##                 test_writer.add_summary(sm, en)
                print('init dev loss', ddloss)
                opstring = opstring + str(ddloss) + ","
                print("alphas=", a)
                print("thetas=",t)
                unique, counts = np.unique(pl, return_counts=True)
                print("Unique,counts of marginals", dict(zip(unique, counts)))
                print("acc",accuracy_score(gold_labels_dev,pl))
                dev_results = precision_recall_fscore_support(np.array(gold_labels_dev),np.array(pl),average="binary")
                print('devresults ', dev_results)
                roc_dev = roc_auc_score(np.array(gold_labels_dev), np.array(m[1,:,0]))
                print('roc score', roc_dev)
                opstring = str(en) + ',' #do not print losses
                opstring = opstring + str(dev_results[2]) + "," + str(roc_dev) + ","
#                 if(dev_results[2]>dev_high[2]):
                dev_high = np.array(dev_results)
#                 dev_high = max(dev_high, dev_results[2])
                print()
                print("test set")
                sess.run(test_init_op)
                a,t,m,pl = sess.run([alphas,thetas,marginals,predict])
                unique, counts = np.unique(pl, return_counts=True)
                print(dict(zip(unique, counts)))
                print("acc",accuracy_score(gold_labels_test,pl))
                test_results = precision_recall_fscore_support(np.array(gold_labels_test),np.array(pl),average="binary")
                print('testresults ', test_results)
                roc_test = roc_auc_score(np.array(gold_labels_test), np.array(m[1,:,0]))
                print('roc score', roc_test)
                opstring = opstring + str(test_results[2]) + "," + str(roc_test)
#                 if(test_results[2]>test_high[2]):
                test_high = np.array(test_results)
                print()
        
                sys.stdout = stdout_saved
                print(opstring)
                print(a, t)
# #                 newptheta = sess.run(marginals)
#                 print("new z ", ztemp)
#                 print("new z shape ", np.array(ztemp).shape)
#                 print('shape ',np.array(newptheta).shape)
                # sys.stdout = open("trash", "w")
                

    return pl, m, dev_high, opstring#test_high

In [9]:
#added 2
r42d = []
r42t = []


a = 0.80

ua = np.repeat(a, len(LF_l))

stdout_saved = orig_stdout
# for rx in [1, len(LF_l), len(train_L_S)]:
# for i in np.linspace(1,1,1):
b = 32


print(discr)
for pen_i in [151]: # [101,102,103,104,105,106,111,112,113,114,115,116]:#[0, 4, 42]:
# for b in [32]:#[32,64,128,512,1024,2048,4096]:
    print(int(pen_i)//10 % 10)
    devmax = np.zeros(3)
    testmax = np.zeros(3)
    for gg in range(1):#[1]:#np.linspace(0,100,101):#, 10, 30, 50, 100]:
#         print("batch-size:",b, "gamma:",gg)
        print('============', pen_i)
        # sys.stdout = open('trash', 'w')
#        sys.stdout = stdout_saved
        _, _, _, ops = train2(1/len(train_L_S),200,batch_size = b, th = tf.truncated_normal_initializer(1,0,seed),\
                                    af = tf.truncated_normal_initializer(0,0.1,seed),\
                                    user_a = ua, \
                                    LF_acc = np.repeat(0.4, len(LF_l)) ,pcl=np.array([-1,1],dtype=np.float64),\
                                    norm=True,smooth=True,penalty=pen_i, Gamma=gg, debug=False,\
                                    a_pr = np.repeat(0.750, len(LF_l)), \
                                           n_pr = np.repeat(100, len(LF_l)),\
                                     isdiscrete = dscr, alpha_max=np.repeat(0.9, len(LF_l)),
                             r_pr = np.repeat(0.750, len(LF_l)))
        sys.stdout = orig_stdout
# get_LF_acc(dev_L_S,gold_labels_dev)
#                 if(devmax_t[2]>devmax[2]):
#         devmax = np.array(devmax_t)
#                 if(testmax_t[2]>testmax[2]):
#         testmax = np.array(testmax_t)
#         r42d.append(devmax)
#         r42t.append(testmax)
#         print("\n\nfinal_result_line: ","thetas:",0,0.1,"batch size",b,"penalty",pen_i,)
##        sys.stdout = stdout_saved
#         print("\n\nfinal_result_line: ","thetas:",0,0.1,"batch size",b,"penalty",pen_i,)
# sys.stdout = open('opfile2', 'a')
# print('rate = 1/train_len')      
# sys.stdout = stdout_saved

[1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
5
============ 151
Using binomial precision constraints
new model smooth with constraints  151
new model smooth with constraints  151
0 loss= 7385.17612725  precision loss= 11787222.0268
ploss 11787222.0268
ratio 0.0
dev set
init dev loss 16795.5552638
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[ 0.96881327  0.96881324  0.96881327  0.96881337  0.96881306  0.96878867
   0.96881312  0.96878931  0.96878947  0.96878815]]
Unique,counts of marginals {0: 2456, 1: 358}
acc 0.885216773276
devresults  (0.31284916201117319, 0.59259259259259256, 0.40950639853747711, None)
roc score 0.788667170572

test set
{0: 2292, 1: 410}
acc 0.879348630644
testresults  (0.36829268292682926, 0.69266055045871555, 0.48089171974522299, None)
roc score 0.822298490153

0,0.409506398537,0.788667170572,0.480891719745,0.822298490153
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  

8 loss= 6732.64110866  precision loss= 10571492.8135
ploss 10571492.8135
ratio 0.0
dev set
init dev loss 15066.7127102
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[ 0.72352635  0.72352612  0.72352453  0.72352478  0.72352584  0.72163695
   0.72352631  0.72164022  0.72164077  0.72164284]]
Unique,counts of marginals {0: 2456, 1: 358}
acc 0.885216773276
devresults  (0.31284916201117319, 0.59259259259259256, 0.40950639853747711, None)
roc score 0.789302091207

test set
{0: 2292, 1: 410}
acc 0.879348630644
testresults  (0.36829268292682926, 0.69266055045871555, 0.48089171974522299, None)
roc score 0.820442575603

8,0.409506398537,0.789302091207,0.480891719745,0.820442575603
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[ 0.72352635  0.72352612  0.72352453  0.72352478  0.72352584  0.72163695
   0.72352631  0.72164022  0.7

17 loss= 6192.19138591  precision loss= 9412770.743
ploss 9412770.743
ratio 0.0
dev set
init dev loss 13421.5031992
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[ 0.4513896   0.4513891   0.45138457  0.45138516  0.45138874  0.44390842
   0.4513898   0.44391917  0.44392042  0.44392809]]
Unique,counts of marginals {0: 2456, 1: 358}
acc 0.885216773276
devresults  (0.31284916201117319, 0.59259259259259256, 0.40950639853747711, None)
roc score 0.789292013102

test set
{0: 2292, 1: 410}
acc 0.879348630644
testresults  (0.36829268292682926, 0.69266055045871555, 0.48089171974522299, None)
roc score 0.820442575603

17,0.409506398537,0.789292013102,0.480891719745,0.820442575603
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[ 0.4513896   0.4513891   0.45138457  0.45138516  0.45138874  0.44390842
   0.4513898   0.44391917  0.443

26 loss= 5880.674764  precision loss= 8486211.02654
ploss 8486211.02654
ratio 0.0
dev set
init dev loss 12110.759002
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[ 0.18349711  0.18349629  0.18348751  0.18348876  0.18349578  0.16562313
   0.18349751  0.16564528  0.16564742  0.16566126]]
Unique,counts of marginals {0: 2456, 1: 358}
acc 0.885216773276
devresults  (0.31284916201117319, 0.59259259259259256, 0.40950639853747711, None)
roc score 0.789292013102

test set
{0: 2292, 1: 410}
acc 0.879348630644
testresults  (0.36829268292682926, 0.69266055045871555, 0.48089171974522299, None)
roc score 0.820442575603

26,0.409506398537,0.789292013102,0.480891719745,0.820442575603
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[ 0.18349711  0.18349629  0.18348751  0.18348876  0.18349578  0.16562313
   0.18349751  0.16564528  0.16

35 loss= 5808.04176603  precision loss= 7799487.53878
ploss 7799487.53878
ratio 0.0
dev set
init dev loss 11144.4803861
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.07758438 -0.07758573 -0.07760275 -0.07760006 -0.07758658 -0.11246886
  -0.07758376 -0.11243266 -0.11242946 -0.11240879]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222507432603

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

35,0.0575539568345,0.222507432603,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.07758438 -0.07758573 -0.07760275 -0.07760006 -0.07758658 -0.11246886
  -0.07758376 -0.1124326

43 loss= 5923.67832458  precision loss= 7384763.30566
ploss 7384763.30566
ratio 0.0
dev set
init dev loss 10565.5026339
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.29882758 -0.29883022 -0.29886751 -0.2988613  -0.29883189 -0.35821109
  -0.29882639 -0.3581605  -0.35815616 -0.35812813]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222509448224

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

43,0.0575539568345,0.222509448224,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.29882758 -0.29883022 -0.29886751 -0.2988613  -0.29883189 -0.35821109
  -0.29882639 -0.3581605

51 loss= 6145.3892705  precision loss= 7134864.57111
ploss 7134864.57111
ratio 0.0
dev set
init dev loss 10220.5600061
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.49477137 -0.49477994 -0.49490639 -0.49488549 -0.49478489 -0.59985553
  -0.4947669  -0.59978685 -0.5997809  -0.59974184]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222517510708

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180816860938

51,0.0575539568345,0.222517510708,0.0537798072045,0.180816860938
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.49477137 -0.49477994 -0.49490639 -0.49488549 -0.49478489 -0.59985553
  -0.4947669  -0.59978685

59 loss= 6351.11704239  precision loss= 7009694.42458
ploss 7009694.42458
ratio 0.0
dev set
init dev loss 10050.0737971
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.63440081 -0.63443541 -0.63496705 -0.63488379 -0.63445159 -0.828044
  -0.63437794 -0.82793843 -0.82792864 -0.82786225]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222582010582

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180816860938

59,0.0575539568345,0.222582010582,0.0537798072045,0.180816860938
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.63440081 -0.63443541 -0.63496705 -0.63488379 -0.63445159 -0.828044
  -0.63437794 -0.82793843 -0

67 loss= 6531.1696367  precision loss= 6965289.51227
ploss 6965289.51227
ratio 0.0
dev set
init dev loss 9992.33098278
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.73391376 -0.7339445  -0.73482502 -0.73470556 -0.73395525 -1.02079923
  -0.73386915 -1.02052448 -1.02049586 -1.020292  ]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222517510708

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180816860938

67,0.0575539568345,0.222517510708,0.0537798072045,0.180816860938
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.73391376 -0.7339445  -0.73482502 -0.73470556 -0.73395525 -1.02079923
  -0.73386915 -1.02052448

75 loss= 6601.40790459  precision loss= 6962177.62278
ploss 6962177.62278
ratio 0.0
dev set
init dev loss 9989.16384881
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.7633021  -0.76331692 -0.76431855 -0.76424899 -0.76334279 -1.07902129
  -0.76336498 -1.078082   -1.07798249 -1.07721619]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222535651298

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180694979982

75,0.0575539568345,0.222535651298,0.0537798072045,0.180694979982
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.7633021  -0.76331692 -0.76431855 -0.76424899 -0.76334279 -1.07902129
  -0.76336498 -1.078082 

83 loss= 6601.41116575  precision loss= 6962177.62237
ploss 6962177.62237
ratio 0.0
dev set
init dev loss 9989.16386186
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.76330753 -0.76331267 -0.7643166  -0.76425121 -0.76333459 -1.07911098
  -0.76331092 -1.07800074 -1.07788542 -1.07720696]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222529604434

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

83,0.0575539568345,0.222529604434,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.76330753 -0.76331267 -0.7643166  -0.76425121 -0.76333459 -1.07911098
  -0.76331092 -1.0780007

91 loss= 6601.4103917  precision loss= 6962177.62512
ploss 6962177.62512
ratio 0.0
dev set
init dev loss 9989.16386721
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.76329257 -0.76329982 -0.76430908 -0.76424316 -0.7633222  -1.07906517
  -0.76329841 -1.07799834 -1.07788303 -1.0771452 ]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222529604434

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

91,0.0575539568345,0.222529604434,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.76329257 -0.76329982 -0.76430908 -0.76424316 -0.7633222  -1.07906517
  -0.76329841 -1.07799834

99 loss= 6601.41039244  precision loss= 6962177.62515
ploss 6962177.62515
ratio 0.0
dev set
init dev loss 9989.16386725
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799847 -1.07788317 -1.07714418]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222529604434

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

99,0.0575539568345,0.222529604434,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.0779984

107 loss= 6601.41039245  precision loss= 6962177.62515
ploss 6962177.62515
ratio 0.0
dev set
init dev loss 9989.16386725
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799847 -1.07788317 -1.07714417]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222529604434

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

107,0.0575539568345,0.222529604434,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799

115 loss= 6601.41039245  precision loss= 6962177.62515
ploss 6962177.62515
ratio 0.0
dev set
init dev loss 9989.16386725
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799847 -1.07788317 -1.07714417]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222529604434

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

115,0.0575539568345,0.222529604434,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799

123 loss= 6601.41039245  precision loss= 6962177.62515
ploss 6962177.62515
ratio 0.0
dev set
init dev loss 9989.16386725
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799847 -1.07788317 -1.07714417]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222529604434

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

123,0.0575539568345,0.222529604434,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799

131 loss= 6601.41039245  precision loss= 6962177.62515
ploss 6962177.62515
ratio 0.0
dev set
init dev loss 9989.16386725
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799847 -1.07788317 -1.07714417]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222529604434

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

131,0.0575539568345,0.222529604434,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799

139 loss= 6601.41039245  precision loss= 6962177.62515
ploss 6962177.62515
ratio 0.0
dev set
init dev loss 9989.16386725
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799847 -1.07788317 -1.07714417]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222529604434

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

139,0.0575539568345,0.222529604434,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799

147 loss= 6601.41039245  precision loss= 6962177.62515
ploss 6962177.62515
ratio 0.0
dev set
init dev loss 9989.16386725
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799847 -1.07788317 -1.07714417]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222529604434

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

147,0.0575539568345,0.222529604434,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799

155 loss= 6601.41039245  precision loss= 6962177.62515
ploss 6962177.62515
ratio 0.0
dev set
init dev loss 9989.16386725
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799847 -1.07788317 -1.07714417]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222529604434

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

155,0.0575539568345,0.222529604434,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799

163 loss= 6601.41039245  precision loss= 6962177.62515
ploss 6962177.62515
ratio 0.0
dev set
init dev loss 9989.16386725
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799847 -1.07788317 -1.07714417]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222529604434

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

163,0.0575539568345,0.222529604434,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799

171 loss= 6601.41039245  precision loss= 6962177.62515
ploss 6962177.62515
ratio 0.0
dev set
init dev loss 9989.16386725
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799847 -1.07788317 -1.07714417]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222529604434

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

171,0.0575539568345,0.222529604434,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799

179 loss= 6601.41039245  precision loss= 6962177.62515
ploss 6962177.62515
ratio 0.0
dev set
init dev loss 9989.16386725
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799847 -1.07788317 -1.07714417]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222529604434

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

179,0.0575539568345,0.222529604434,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799

187 loss= 6601.41039245  precision loss= 6962177.62515
ploss 6962177.62515
ratio 0.0
dev set
init dev loss 9989.16386725
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799847 -1.07788317 -1.07714417]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222529604434

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

187,0.0575539568345,0.222529604434,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799

195 loss= 6601.41039245  precision loss= 6962177.62515
ploss 6962177.62515
ratio 0.0
dev set
init dev loss 9989.16386725
alphas= [ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ]
thetas= [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799847 -1.07788317 -1.07714417]]
Unique,counts of marginals {0: 918, 1: 1896}
acc 0.30170575693
devresults  (0.031645569620253167, 0.31746031746031744, 0.057553956834532377, None)
roc score 0.222529604434

test set
{0: 949, 1: 1753}
acc 0.30977054034
testresults  (0.030233884768967486, 0.24311926605504589, 0.053779807204464743, None)
roc score 0.180818707619

195,0.0575539568345,0.222529604434,0.0537798072045,0.180818707619
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[-0.76329217 -0.76329955 -0.76430893 -0.76424301 -0.76332195 -1.07906499
  -0.76329823 -1.07799

1,0.43006263048,0.762761400857,0.498327759197,0.818907060231
[ 0.11754463 -0.18578547  0.06448096  0.05554933  0.10637787  0.04133855
  0.22405651 -0.02316195  0.09187082  0.01032644] [[ 0.94514111  0.94514112  0.94514827  0.94514484  0.94514112  1.05462089
   0.9452137   1.05444394  1.05452913  1.05446836]]
   
   Cont: 111
   10,0.407002188184,0.746422776518,0.483636363636,0.80223891622
[ 0.11754463 -0.18578547  0.1191033   0.11100223  0.10637787  0.04133855
  0.26315504 -0.05675751  0.04734599 -0.02641954] [[ 0.90034616  0.90035037  0.89118076  0.89039836  0.90035211  1.12112825
   0.90581917  1.08926482  1.09932218  1.09215784]]
   
   Cont: 112
   9,0.296052631579,0.685242630385,0.375,0.737930092039
[ 0.11754463 -0.18578547  0.16593194  0.15628545  0.10637787  0.04133855
  0.32336718 -0.11055429 -0.00683302 -0.08423155] [[ 0.84517908  0.84518043  0.85065344  0.85047745  0.84518095  1.16100097
   0.85390063  1.16385916  1.16524347  1.16419562]]

    Cont: 112 with max_alpha = 0.7
    140,0.414587332054,0.780391030486,0.481300813008,0.830555924892
[ 1.11754463  0.81421453  1.00955965  1.00062431  1.10637787  1.04133855
  1.169206    1.03120785  1.14633471  1.0647233 ] [[ 0.2137293   0.22307502  4.16844262  3.91339975  0.23617798  1.17968475
   0.48896712  2.30967195  2.14310226 -0.46482738]]
   
   Discrete: 111
   10,0.428274428274,0.777492567397,0.49569707401,0.819183139062
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[ 0.82293943  0.82294084  0.82296135  0.82295826  0.82294152  1.18440805
   0.82293851  1.18439077  1.18438875  1.18437385]]
   
   Discrete: 112
   10,0.428274428274,0.777492567397,0.49569707401,0.819183139062
[ 0.11754463 -0.18578547  0.00955965  0.00062431  0.10637787  0.04133855
  0.169206    0.03120785  0.14633471  0.0647233 ] [[ 0.82293943  0.82294084  0.82296135  0.82295826  0.82294152  1.18440805
   0.82293851  1.18439077  1.18438875  1.18437385]]